### ISSUE TITLE
Helper base class?

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#874]
Jon says:

---

I suspect there are lots of times when a helper function (e.g., create()) would like to invoke the default behavior (e.g., via this.inherited()) and then add a couple of other things. The way our helpers are today, helper functions are mostly substitutes for the standard behavior.

Also, by having a base class, we can include JSDoc comments in the base class that provides full documentation of the various helper functions.

Sometime afte

---

Javier responds:

---

So, I see two ways of handling this:

1) (Jon's suggestion) Create a base class for Helpers, CreateTools, etc, but allow widget libraries to override the base implementation with their own impl.

2) The Helper, CreateTool, etc impls are completely self contained but may contain a return value.  The default implementation would call the Helper functions, if available, at some point in the code.  The base implementation may even decide to do something different based on the return value from the Helper functions.

The major difference between the two implementations is one of separation and control.  With (1), the Helper and CreatTool objects are elevated to the same level as the core Maqetta code, and in fact have a bit more control, since they can determine whether to call the default implementation in the core.  With (2), control is kept in the core, and the Helper objects only have access to the data necessary to perform their job.

When architecting the Library APIs, I've been approaching it from the point of (2).  I think this makes more sense.

---

**Comment by jhpedemonte** [comment]
Continuing my thinking:

By using callbacks instead of OO, we can wrap the callbacks in **try/catch blocks**, so errors in the library code don't impact the core code.

Callbacks would generally be invoked at the beginning of the core function.  If the callback returns **false**, then the default behavior is prevented.

If the callback returns **true or a continuation**, then the default behavior is executed.  A continuation function is invoked after the default behavior (again, within a try/catch).

Just thinking out loud, so this isn't all fully formed...

**Comment by JonFerraiolo** [comment]
@peller, @billreed63, @childsb, @waynevicknair

In the absence of OO approaches, I actually used the "continuation" approach in at least one place in the code: VisualEditorOutline.js:267 When I added this code, I felt pretty dirty about it and sensed that there just had to be a cleaner more elegant approach using OO techniques, but maybe the continuation approach overall ends up being better than other options.

**Comment by billreed63** [comment]
If we are voting I vote for the OO approach, option 1

**Comment by jhpedemonte** [comment]
The "continuation" approach doesn't have to look ugly, and in fact, is more in keeping with how JavaScript was designed.  And frankly, our use of OO in some parts of the code is pretty ugly.

We can accomplish what we need to with either method.  But as I stated, the callback/continuation approach gives more control to the core Maqetta code, over the library code.

**Comment by peller** [comment]
I've become more convinced that we shouldn't even have helper constructors at all.  There's no correlation between helper instance and widget instance, so it has relatively little meaning to use the 'this' pointer as a context to store state.  We can achieve inheritance by simply using mixins on objects.

**Comment by JonFerraiolo** [comment]
Yes, agreed, no reason for helper instances, and probably would be nice and clean to mixin the helpers into the _dvWidget objects.

**Comment by jhpedemonte** [comment]
-> M6

**Comment by jhpedemonte** [comment]
-> M7
